src/lib/util/hash.cpp

```diff
@@ -27,48 +30,80 @@ char const *const hash_collection::HASH_TYPES_CRC_SHA1 = "RS";
 char const *const hash_collection::HASH_TYPES_ALL = "RS";
 
 
-
 //**************************************************************************
-//  HASH COLLECTION
+//  HASH CREATOR
 //**************************************************************************
 
-//-------------------------------------------------
-//  hash_collection - constructor
-//-------------------------------------------------
+namespace {
 
-hash_collection::hash_collection()
-	: m_has_crc32(false),
-		m_has_sha1(false),
-		m_creator(nullptr)
+class hash_creator : public write_stream
 {
-}
+public:
+	// constructor
+	hash_creator(hash_collection &hashes, bool doing_crc32, bool doing_sha1)
+		: m_hashes(hashes)
+	{
+		if (doing_crc32)
+			m_crc32_creator = std::make_optional<crc32_creator>();
+		if (doing_sha1)
+			m_sha1_creator = std::make_optional<sha1_creator>();
```

Comment:
Why the extra copies?  You can use `.emplace()` on `std::optional`.

Reply:
The opposite is preferred for `std::unique_ptr`. Oh, why does C++ have to be so inconsistent about everything?

Reply:
Not by me.  I’m perfectly fine with people doing `foo.reset(new x)` on a `std::unique_ptr`.  I don’t subscribe to the religion of always using `std::make_unique`.

The issue is with code like this:
```
foo(std::unique_ptr<C>(new C), bar());
```
There’s no sequence point between the argument expressions, so it’s possible for the compiler to realise it as:
* `new C`
* `bar()`
* construct `std::unique_ptr<C>`
* Call `foo`

Then if `bar()` throws, you leak a `C`.

This is solved by:
```
foo(std::make_unique<C>(), bar());
```
because the return from `std::make_unique` generates a sequence point that guarantees the `unique_ptr` will always be constructed immediately after allocating the `C` instance, so you don’t get the leak.